### PR TITLE
Fix #107: Isolate business logic of Data Adapter into separate service

### LINKED ISSUE #107: Isolate business logic of Data Adapter into separate service
We currently have business logic spread across multiple controllers. We should introduce an interface, let customers implement it according to their business logic, and autowire the service in controllers.

### PR DESCRIPTION
**Pull request by romanstrobl**
Reimplemented with a better interface for consumers.

### CODE REVIEW

File: powerauth-data-adapter/src/main/java/io/getlime/security/powerauth/lib/dataadapter/api/DataAdapter.java [hunk 1]
```diff
@@ -34,55 +35,50 @@
...
      *
-     * @param authenticationRequest Authentication request.
-     * @return Authentication response.
-     * @throws MethodArgumentNotValidException Thrown when input validation fails.
+     * @param username Username for user authentication.
+     * @param password Password for user authentication.
      * @throws AuthenticationFailedException Thrown when authentication fails.
      */
-    AuthenticationResponse authenticateUser(AuthenticationRequest authenticationRequest) throws MethodArgumentNotValidException, AuthenticationFailedException;
+    void authenticateUser(String username, String password) throws AuthenticationFailedException;
```
Comment (petrdvorak) [inline powerauth-data-adapter/src/main/java/io/getlime/security/powerauth/lib/dataadapter/api/DataAdapter.java]: We should return something from this method, for example an object with user information (at least, user ID has to be returned). It feels weird that we only let the method go and let it fail on authentication failure...

File: powerauth-data-adapter/src/main/java/io/getlime/security/powerauth/lib/dataadapter/controller/AuthenticationController.java [hunk 2]
```diff
@@ -58,7 +63,18 @@ public AuthenticationController(DataAdapter dataAdapter) {
...
+        AuthenticationRequestValidator validator = new AuthenticationRequestValidator();
+        BeanPropertyBindingResult result = new BeanPropertyBindingResult(authenticationRequest, "authenticationRequest");
+        ValidationUtils.invokeValidator(validator, authenticationRequest, result);
+        if (result.hasErrors()) {
+            // getEnclosingMethod() on new object returns a reference to current method
+            MethodParameter methodParam = new MethodParameter(new Object(){}.getClass().getEnclosingMethod(),0);
+            throw new MethodArgumentNotValidException(methodParam, result);
+        }
+        dataAdapter.authenticateUser(authenticationRequest.getUsername(), authenticationRequest.getPassword());
+        AuthenticationResponse response = new AuthenticationResponse(authenticationRequest.getUsername());
```
Comment (petrdvorak) [inline powerauth-data-adapter/src/main/java/io/getlime/security/powerauth/lib/dataadapter/controller/AuthenticationController.java]: This is the issue - "username" is not "userId". Generally, these values differ. While username in a typical bank is either a client number, e-mail or username, it always translates to canonical user ID...

File: powerauth-data-adapter/src/main/java/io/getlime/security/powerauth/lib/dataadapter/controller/BankAccountController.java [hunk 3]
```diff
@@ -51,9 +52,9 @@ public BankAccountController(DataAdapter dataAdapter) {
      * @return Response with user details.
      */
     @RequestMapping(value = "/list", method = RequestMethod.POST)
-    public @ResponseBody ObjectResponse<BankAccountListResponse> fetchBankAccounts(@RequestBody ObjectRequest<BankAccountListRequest> request) throws MethodArgumentNotValidException {
+    public @ResponseBody ObjectResponse<BankAccountListResponse> fetchBankAccounts(@RequestBody ObjectRequest<BankAccountListRequest> request) throws MethodArgumentNotValidException, UserNotFoundException {
         BankAccountListRequest bankAccountListRequest = request.getRequestObject();
-        BankAccountListResponse response = dataAdapter.fetchBankAccounts(bankAccountListRequest);
+        BankAccountListResponse response = new BankAccountListResponse(bankAccountListRequest.getUserId(), dataAdapter.fetchBankAccounts(bankAccountListRequest.getUserId()));
```
Comment (petrdvorak) [inline powerauth-data-adapter/src/main/java/io/getlime/security/powerauth/lib/dataadapter/controller/BankAccountController.java]: This code could use a little polist... for example:

```java
String userId = bankAccountListRequest.getUserId();
// ...
```

Generally, I prefer slightly more lines of code with a clear meaning, than large method parameter chaining...

File: powerauth-data-adapter/src/main/java/io/getlime/security/powerauth/lib/dataadapter/controller/SMSAuthorizationController.java [hunk 4]
```diff
@@ -54,8 +64,21 @@ public SMSAuthorizationController(DataAdapter dataAdapter) {
      */
     @RequestMapping(value = "create", method = RequestMethod.POST)
     public @ResponseBody ObjectResponse<CreateSMSAuthorizationResponse> create(@RequestBody ObjectRequest<CreateSMSAuthorizationRequest> request) throws MethodArgumentNotValidException, SMSAuthorizationFailedException {
-        CreateSMSAuthorizationRequest createSMSAuthorizationRequest = request.getRequestObject();
-        CreateSMSAuthorizationResponse response = dataAdapter.createAuthorizationSMS(createSMSAuthorizationRequest);
+        CreateSMSAuthorizationRequest smsRequest = request.getRequestObject();
+        // input validation is handled by CreateSMSAuthorizationRequestValidator
+        // validation is invoked manually because of the generified Request object
+        BeanPropertyBindingResult result = new BeanPropertyBindingResult(smsRequest, "createSMSAuthorizationRequest");
```
Comment (petrdvorak) [inline powerauth-data-adapter/src/main/java/io/getlime/security/powerauth/lib/dataadapter/controller/SMSAuthorizationController.java]: I do not understand this code - do we really need to use custom binding?

File: powerauth-data-adapter/src/main/java/io/getlime/security/powerauth/lib/dataadapter/controller/SMSAuthorizationController.java [hunk 5]
```diff
@@ -54,8 +64,21 @@ public SMSAuthorizationController(DataAdapter dataAdapter) {
...
+        ValidationUtils.invokeValidator(requestValidator, smsRequest, result);
+        if (result.hasErrors()) {
+            // getEnclosingMethod() on new object returns a reference to current method
+            MethodParameter methodParam = new MethodParameter(new Object() {
```
Comment (petrdvorak) [inline powerauth-data-adapter/src/main/java/io/getlime/security/powerauth/lib/dataadapter/controller/SMSAuthorizationController.java]: This also seems like something we should not need in the code.

File: powerauth-data-adapter/src/main/java/io/getlime/security/powerauth/lib/dataadapter/impl/service/DataAdapterService.java [hunk 6]
```diff
@@ -39,47 +22,23 @@
...
-            MethodParameter methodParam = new MethodParameter(new Object(){}.getClass().getEnclosingMethod(),0);
-            throw new MethodArgumentNotValidException(methodParam, result);
-        }
-
-        // here will be the real authentication - call to the backend providing authentication
-        if ("test".equals(authenticationRequest.getPassword())) {
-            return new AuthenticationResponse(authenticationRequest.getUsername());
-        } else {
-            throw new AuthenticationFailedException("login.authenticationFailed");
+    public void authenticateUser(String username, String password) throws AuthenticationFailedException {
```
Comment (petrdvorak) [inline powerauth-data-adapter/src/main/java/io/getlime/security/powerauth/lib/dataadapter/impl/service/DataAdapterService.java]: dtto as above ...

**Comment by romanstrobl** [comment]
I have resolved the comments about return value from user authentication call and avoiding calling methods in method arguments.

The validation code is a bit more tricky. I replaced manual Validator invocation by injecting specific Validators into controllers and doing validation on the ObjectRequests. Luckily Spring supports field nesting and it works with generified ObjectRequest, so this code works to my surprise:

```
ValidationUtils.rejectIfEmptyOrWhitespace(errors, "requestObject.username", "login.username.empty");
```

The error handling for failed validation is still a bit cumbersome (throwing the MethodArgumentNotValidException), we can look for a cleaner solution, however current code should work.

**Pull request merged**